### ISSUE TITLE
Schedule fails at large size

### ISSUE THREAD

**Issue by liangyuRain** [microsoft/msccl#38]
Hi, we have observed that the following allreduce schedule `test.txt` fails only at large size (16777216B) with branch `msccl/merging_with_nccl`. We observed this bug in our testbed when there are large number of nodes (8+), with large `nchunksperloop`, and at large sizes. The other schedule `test2.txt` attached is the same schedule but with smaller `nchunksperloop`. `test2.txt` succeeds the datacheck.

[test.txt](https://github.com/microsoft/msccl/files/9147483/test.txt)
[test2.txt](https://github.com/microsoft/msccl/files/9147484/test2.txt)

We have modify the `coll` of `test.txt` to test with `all_reduce_perf`, `reduce_scatter_perf`, and `all_gather_perf`. The results are as followed:

### `all_reduce_perf`
```
#
#                                                       out-of-place                       in-place          
#       size         count      type   redop     time   algbw   busbw  error     time   algbw   busbw  error
#        (B)    (elements)                       (us)  (GB/s)  (GB/s)            (us)  (GB/s)  (GB/s)       
        8192          2048     float     sum    39.32    0.21    0.36  2e-07
       16384          4096     float     sum    39.02    0.42    0.73  2e-07
       32768          8192     float     sum    39.24    0.84    1.46  2e-07
       65536         16384     float     sum    39.54    1.66    2.90  2e-07
      131072         32768     float     sum    41.66    3.15    5.51  5e-07
      262144         65536     float     sum    45.30    5.79   10.13  5e-07
      524288        131072     float     sum    53.41    9.82   17.18  5e-07
     1048576        262144     float     sum    71.83   14.60   25.55  5e-07
     2097152        524288     float     sum    112.6   18.63   32.61  5e-07
     4194304       1048576     float     sum    175.7   23.87   41.77  5e-07
     8388608       2097152     float     sum    314.7   26.66   46.65  5e-07
    16777216       4194304     float     sum    600.5   27.94   48.89  2e+00
# Out of bounds values : 2 FAILED
# Avg bus bandwidth    : 19.4778 
#
```

### `reduce_scatter_perf`
```
#
#                                                       out-of-place                       in-place          
#       size         count      type   redop     time   algbw   busbw  error     time   algbw   busbw  error
#        (B)    (elements)                       (us)  (GB/s)  (GB/s)            (us)  (GB/s)  (GB/s)       
        8192           256     float     sum    39.58    0.21    0.18  2e-07
       16384           512     float     sum    38.75    0.42    0.37  2e-07
       32768          1024     float     sum    39.41    0.83    0.73  2e-07
       65536          2048     float     sum    39.75    1.65    1.44  2e-07
      131072          4096     float     sum    41.78    3.14    2.74  5e-07
      262144          8192     float     sum    44.95    5.83    5.10  5e-07
      524288         16384     float     sum    53.11    9.87    8.64  5e-07
     1048576         32768     float     sum    72.23   14.52   12.70  5e-07
     2097152         65536     float     sum    112.1   18.71   16.37  5e-07
     4194304        131072     float     sum    201.5   20.81   18.21  5e-07
     8388608        262144     float     sum    330.6   25.37   22.20  5e-07
    16777216        524288     float     sum    602.1   27.87   24.38  5e-07
# Out of bounds values : 0 OK
# Avg bus bandwidth    : 9.42284 
#
```

### `all_gather_perf`
```
g1lmd6: Test CUDA failure common.cu:506 'an illegal memory access was encountered'
 .. g1lmd6 pid 1269293: Test failure common.cu:607
 .. g1lmd6 pid 1269293: Test failure common.cu:798
 .. g1lmd6 pid 1269293: Test failure all_gather.cu:93
 .. g1lmd6 pid 1269293: Test failure common.cu:824
 .. g1lmd6 pid 1269293: Test failure common.cu:1199
 .. g1lmd6 pid 1269293: Test failure common.cu:1039
```

We also tried to change `test.txt` to `Simple` and `LL128` protocols. `LL128` is fine, but a `divide-by-zero` error is observed with `Simple`. `test2.txt` works perfectly with any protocol.

In summary, we observed the following bugs:
- `test.txt` fails datacheck at large size.
- `test.txt` cannot run with `all_gather_perf`.
- `test.txt` has `divide-by-zero` error when changed to `Simple` protocol.

**Comment by saeedmaleki** [comment]
Integer overflow problem!
1st and 3rd issue from your summary are fixed. Will investigate the middle issue.
Please pull and confirm.

P.S.: When you have such large send counts, you will run into deadlock issues wrt number of slots available for each protocol. 2 for Simple and 8 for LL/LL128.

**Comment by liangyuRain** [comment]
Hi, @saeedmaleki. Thank you for the bug fix. However, I still observe the first bug with `test.txt` and latest `origin/merging_with_nccl`.
```
#
#                                                       out-of-place                       in-place          
#       size         count      type   redop     time   algbw   busbw  error     time   algbw   busbw  error
#        (B)    (elements)                       (us)  (GB/s)  (GB/s)            (us)  (GB/s)  (GB/s)       
        8192          2048     float     sum    38.95    0.21    0.37  2e-07
       16384          4096     float     sum    39.15    0.42    0.73  2e-07
       32768          8192     float     sum    38.98    0.84    1.47  2e-07
       65536         16384     float     sum    39.57    1.66    2.90  2e-07
      131072         32768     float     sum    41.42    3.16    5.54  5e-07
      262144         65536     float     sum    45.05    5.82   10.18  5e-07
      524288        131072     float     sum    52.96    9.90   17.32  5e-07
     1048576        262144     float     sum    71.47   14.67   25.67  5e-07
     2097152        524288     float     sum    106.1   19.77   34.60  5e-07
     4194304       1048576     float     sum    172.4   24.33   42.58  5e-07
     8388608       2097152     float     sum    330.1   25.42   44.48  5e-07
    16777216       4194304     float     sum    600.5   27.94   48.89  2e+00
# Out of bounds values : 1 FAILED
# Avg bus bandwidth    : 19.5615 
#
```

**Comment by liangyuRain** [comment]
For more information, we are testing a bunch of 8+ node schedules on our optical testbed. We have to modify `msccl/merging_with_nccl` to utilize the optical links. We observe that the following 12-node allreduce schedule fails the datacheck at large sizes.

[schedule.txt](https://github.com/microsoft/msccl/files/9154655/schedule.txt)


The 1st column is size. The 3rd column is the datacheck result.
```
1536	103.8	True
3072	104.2	True
6144	105.6	True
12288	111.1	True
24576	117.4	True
49152	142.6	True
98304	170.3	True
196608	243.5	True
393216	368.7	True
786432	593.5	True
1572864	1239.3	False
3145728	2288.3	False
6291456	4141.8	False
12582912	8170.3	False
25165824	20140.0	False
50331648	33297.0	False
100663296	67656.0	False
201326592	138033.0	False
402653184	277404.0	False
805306368	549966.0	False
1610612736	1139708.0	False
```

However, we are unable to reproduce this bug with original `msccl`, because our Lambda machine has only 8 nodes. We have tested the schedule with modification based on an earlier version of `msccl` (commit `e52c525`) and `master` version. The earlier version works fine but the `master` version fails. Could you please check if you can reproduce this issue with original `msccl`?

**Comment by saeedmaleki** [comment]
Ah, OK, I was using `-d int8` and saw no problem but I could reproduce it with `-d int32`. I am sure there are more int overflow happening since in our use cases `nchunksPerLoop` are never too large. We will investigate the problem. Do you happen to have a smaller schedule we could debug?

**Comment by liangyuRain** [comment]
We were using `float` as our datatype. Unfortunately, we do not have a smaller schedule right now. The following schedule is an allreduce schedule on 12-node ring. It seems to fail intermittently at certain sizes.

[schedule2.txt](https://github.com/microsoft/msccl/files/9158645/schedule2.txt)

```
768	429.2	True
1536	426.8	True
3072	426.6	True
6144	428.2	True
12288	445.1	True
24576	449.5	True
49152	461.3	True
98304	538.4	True
196608	489.8	True
393216	652.0	True
786432	928.6	True
1572864	1433.5	True
3145728	2506.2	True
6291456	4568.1	True
12582912	9237.9	True
25165824	19046.0	True
50331648	37777.0	True
100663296	74248.0	True
201326592	151364.0	True
402653184	327236.0	False
805306368	595592.0	True
1610612736	1194389.0	False
```

I do want to mention that the `schedule.txt` attached earlier does not fail at an earlier version of msccl (commit `e52c525`).

**Comment by saeedmaleki** [comment]
Found the problem with the allgather. With allgather collective, input is `nranks` times smaller than output even though input and output are aliasing with in-place allgather. Therefore, `inputPointer == outputPointers + rank * size/nranks` and when your offset from the `inputBuffer` is larger than `nchunksperloop` you get segfaults. So, if you change your input buffers in the schedule to output buffer (the larger buffer), allgather won't segfault. I will add a checker to avoid such a scenario and MSCCL would give a warning.

About the correctness bug... We still cannot figure out what's going on as it is a non-deterministic one. Can you please try out two things and see if they are true?
- Is the correctness a non-deterministic one? Please run the example a few times back to back and see if they all fail.
- Can you please try increasing the `NCCL_LL_BUFFSIZE` and see if it fixes your issue? The default value is 256KB. Try 1MB instead.

**Comment by liangyuRain** [comment]
Hi, @saeedmaleki. Thank you for the suggestions. Unfortunately, the testbed is not available right now. I will try them once it becomes available.

**Comment by liangyuRain** [comment]
I just had a few runs on the testbed. Here are the answers:
- The correctness error does seem to be a non-deterministic one. Somehow, I cannot reproduce the error of `schedule2.txt` right now. However, the error of `schedule.txt` is stable.
- Increasing `NCCL_LL_BUFFSIZE` seems to delay the correctness failure of `schedule.txt`. When `NCCL_LL_BUFFSIZE` is default, datacheck failure starts from size 1572864. When set to 1MB, the failure starts from 3145728. When set to 5MB, the failure starts from 100663296.

BTW, our `mpirun` command is as followed:
```python
mpi_command = (
    f'timeout 180s '
    f'mpirun -H {",".join(hosts)} -np {len(servers)} --mca btl self,tcp --mca btl_tcp_if_include enp3s0f0 --mca pml ^ucx --mca mtl ^ofi --mca coll ^hcoll '
    f'-x NCCL_NET="IB" -x NCCL_IB_GID_INDEX=3 -x LD_LIBRARY_PATH={server_base_dir}/submodules/cloptsccl/build/lib: '
    f"-x NCCL_NET_SHARED_BUFFERS=0 -x NCCL_MIN_CHANNELS=4 -x MSCCL_XML_FILES={server_base_dir}/{schedule_file} "
    f"-x NCCL_DEBUG=WARN "
    # f"-x NCCL_LL_BUFFSIZE=33554432 "
    f"-x NCCL_ALGO=MSCCL numactl -m 0 -N 0 {server_base_dir}/submodules/nccl-tests/build/{executable} -b {b} -e {b} -g 1 -f 2 "
    f"-c {1 if check else 0} {'-w 1000 -n 1000' if b <= 2 ** 20 else '-w 5 -n 20'}"
)
```

**Comment by saeedmaleki** [comment]
Hi @liangyuRain, please pull. We believe the problem is fixed now for test.xml! I will add a fix for the allgather problem as well soon. Please let me know if you confront more issues.

**Comment by liangyuRain** [comment]
Hi @saeedmaleki, thanks for the bug fix! I saw `test.txt` and `schedule.txt` are running perfectly right now. However, I have found that multi-channel versions of `schedule.txt` fail at intermediate sizes. The failure at these sizes were also observed before the bug fix, so I believe it is a separate issue. Please let me know if you can reproduce this.

The multi-channel version should be similar to the multi-instance in GC3. The number of channels should be equal to the number of instances.

### 1 channel Simple
[schedule_c1.txt](https://github.com/microsoft/msccl/files/9185690/schedule_c1.txt)
Identical to `schedule.txt` above
Datacheck succeeds at all sizes

### 2 channels Simple
[schedule_c2.txt](https://github.com/microsoft/msccl/files/9185695/schedule_c2.txt)
Datacheck succeeds at all sizes

### 4 channels Simple
[schedule_c4.txt](https://github.com/microsoft/msccl/files/9185696/schedule_c4.txt)
Experiment runs from size 6144 to 1610612736
Datacheck fails from size 3145728 to 6291456

### 8 channels Simple
[schedule_c8.txt](https://github.com/microsoft/msccl/files/9185711/schedule_c8.txt)
Experiment runs from size 12288 to 1610612736
Datacheck fails from size 3145728 to 100663296

### 8 channels LL
Experiment runs from size 12288 to 1610612736
Datacheck fails from size 1572864 to 25165824

**Comment by saeedmaleki** [comment]
I could replicate the 8 channels Simple schedule, but the rest seem to be fine. Let us see if we it is an obvious problem. Meanwhile, if you could reproduce a smaller schedule with 2,4, or 8 GPUs, that would be great.

**Comment by liangyuRain** [comment]
I have tried a bunch of 8-GPU schedules but haven't been able to reproduce the problem. I will continue to try some other schedules.

**Comment by saeedmaleki** [comment]
I think I see an issue with your schedule_c8.txt. The following two instructions on GPU 3 will have a race condition:

Threadblock 6 executes this instruction:
`<step s="1" type="re" srcbuf="s" srcoff="40" dstbuf="i" dstoff="2568" cnt="8" depid="5" deps="0" hasdep="0"/>`
Now threadblock 1 executes this instruction:
`<step s="19" type="cpy" srcbuf="s" srcoff="4272" dstbuf="i" dstoff="2560" cnt="24" depid="-1" deps="-1" hasdep="1"/>`
Then threadblock 6 executes this instruction:
`<step s="2" type="cpy" srcbuf="i" srcoff="2568" dstbuf="s" dstoff="632" cnt="24" depid="-1" deps="-1" hasdep="0"/>`

The order of last two instructions can be swapped since there is no proper guard preventing it and therefore, it causes a data race. The only `hasdep` on threadblock 6 is for step 18 but it seems to be not effective in preventing the above data race. Please correct me if I am wrong.

**Comment by saeedmaleki** [comment]
Also, the guard for `i_chunks` and out of boundary check is added. It forces `i_chunks` with `allgather` to be `nchunksperloop/nRanks` and etc.

**Comment by liangyuRain** [comment]
Hi @saeedmaleki, thank you for the debugging. I believe the second instruction is guaranteed to be executed after the third instruction. The dependency actually involves send/recv through other gpus. In my understanding, if there is a send-recv pair between two gpus, then all instructions after the recv are guaranteed to be executed later than the instructions before the send. The dependency from the second instruction to the third is as followed:
```
<tb=(3, 6)><step s="2" type="cpy" srcbuf="i" srcoff="2568" dstbuf="s" dstoff="632" cnt="24" depid="-1" deps="-1" hasdep="0"/>
<tb=(3, 6)><step s="3" type="cpy" srcbuf="i" srcoff="2832" dstbuf="s" dstoff="656" cnt="16" depid="-1" deps="-1" hasdep="0"/>
<tb=(3, 6)><step s="4" type="s" srcbuf="s" srcoff="632" dstbuf="s" dstoff="552" cnt="40" depid="-1" deps="-1" hasdep="0"/>
<tb=(9, 3)><step s="4" type="r" srcbuf="s" srcoff="632" dstbuf="s" dstoff="552" cnt="40" depid="-1" deps="-1" hasdep="1"/>
<tb=(9, 7)><step s="4" type="re" srcbuf="s" srcoff="532" dstbuf="i" dstoff="2588" cnt="4" depid="3" deps="4" hasdep="0"/>
<tb=(9, 7)><step s="5" type="re" srcbuf="s" srcoff="572" dstbuf="i" dstoff="2588" cnt="4" depid="-1" deps="-1" hasdep="1"/>
<tb=(9, 4)><step s="9" type="s" srcbuf="i" srcoff="2560" dstbuf="s" dstoff="3136" cnt="32" depid="7" deps="5" hasdep="0"/>
<tb=(10, 5)><step s="7" type="r" srcbuf="i" srcoff="2560" dstbuf="s" dstoff="3136" cnt="32" depid="-1" deps="-1" hasdep="1"/>
<tb=(10, 7)><step s="7" type="re" srcbuf="s" srcoff="3100" dstbuf="i" dstoff="2588" cnt="4" depid="5" deps="7" hasdep="0"/>
<tb=(10, 7)><step s="8" type="re" srcbuf="s" srcoff="3132" dstbuf="i" dstoff="2588" cnt="4" depid="-1" deps="-1" hasdep="0"/>
<tb=(10, 7)><step s="9" type="re" srcbuf="s" srcoff="3164" dstbuf="i" dstoff="2588" cnt="4" depid="-1" deps="-1" hasdep="0"/>
<tb=(10, 7)><step s="10" type="re" srcbuf="s" srcoff="3196" dstbuf="i" dstoff="2588" cnt="4" depid="-1" deps="-1" hasdep="1"/>
<tb=(10, 2)><step s="28" type="s" srcbuf="i" srcoff="2560" dstbuf="i" dstoff="2560" cnt="32" depid="7" deps="10" hasdep="0"/>
<tb=(4, 7)><step s="14" type="r" srcbuf="i" srcoff="2560" dstbuf="i" dstoff="2560" cnt="32" depid="-1" deps="-1" hasdep="1"/>
<tb=(4, 0)><step s="29" type="cpy" srcbuf="i" srcoff="2560" dstbuf="s" dstoff="4152" cnt="24" depid="7" deps="14" hasdep="0"/>
<tb=(4, 0)><step s="30" type="s" srcbuf="s" srcoff="4136" dstbuf="s" dstoff="4256" cnt="40" depid="-1" deps="-1" hasdep="0"/>
<tb=(3, 1)><step s="17" type="r" srcbuf="s" srcoff="4136" dstbuf="s" dstoff="4256" cnt="40" depid="-1" deps="-1" hasdep="0"/>
<tb=(3, 1)><step s="18" type="cpy" srcbuf="s" srcoff="4256" dstbuf="i" dstoff="1792" cnt="16" depid="-1" deps="-1" hasdep="0"/>
<tb=(3, 1)><step s="19" type="cpy" srcbuf="s" srcoff="4272" dstbuf="i" dstoff="2560" cnt="24" depid="-1" deps="-1" hasdep="1"/>
```
The `<tb=(a,b)>` prefix means the instruction is on threadblock `b` of gpu `a`.

We have constructed a dependency graph to verify the existence of dependency between any two instructions involving the same chunk. Unless we have misunderstood the implementation of MSCCL, I think it is very unlikely for the schedules to have dependency issue.

**Comment by saeedmaleki** [comment]
Thanks for your data dependence analysis, it is correct. I think I see the issue but can you please check it for us?
Please move the `__syncthreads()` out of the condition in [here](https://github.com/microsoft/msccl/blob/bc540d9a17fb40e3eef773e8898a8797b97013da/src/collectives/device/msccl_interpreter.h#L187) and see if it fixes your issues. If so, we will find a performant solution for it.

**Comment by liangyuRain** [comment]
Hi @saeedmaleki, to make sure, do you mean to move `__syncthreads()` to immediately before the if statement at line 186? In addition, should I make this change on branch `merging_with_nccl` or the commit your link is pointing to?

**Comment by saeedmaleki** [comment]
ah, they are the same, but just to be sure, make the change to `merging_with_nccl`. And yes, immediately before the if statement. From my end, it looks like it fixes all the issues.

**Comment by liangyuRain** [comment]
I have tested both moving `__syncthreads()` and the latest `merging_with_nccl`. The 8-channel version still fails at intermediate sizes. It seems to be a non-deterministic issue.

**Comment by saeedmaleki** [comment]
Hmm, does it fail at the same sizes as before? 3MB to 96MB with Simple? Can you please share the output of nccl-tests?

**Comment by liangyuRain** [comment]
Yes, here are the output and summary

[AR_Simple_c8.txt](https://github.com/microsoft/msccl/files/9193836/AR_Simple_c8.txt)
```
12288	883.9	True
24576	963.7	True
49152	888.3	True
98304	896.4	True
196608	859.7	True
393216	696.2	True
786432	725.4	True
1572864	1031.2	True
3145728	1245.4	False
6291456	1701.8	False
12582912	2825.7	False
25165824	5244.1	False
50331648	10191.0	False
100663296	20300.0	False
201326592	39237.0	True
402653184	79060.0	True
805306368	156566.0	True
1610612736	307361.0	True
```

[AR_LL_c8.txt](https://github.com/microsoft/msccl/files/9193837/AR_LL_c8.txt)
```
12288	406.2	True
24576	462.5	True
49152	484.1	True
98304	484.2	True
196608	492.9	True
393216	666.9	True
786432	870.8	False
1572864	1415.5	False
3145728	2320.9	False
6291456	4254.6	False
12582912	8422.2	False
25165824	16754.0	False
50331648	38001.0	False
100663296	69517.0	True
201326592	139287.0	True
402653184	278598.0	False
805306368	560906.0	True
1610612736	1121097.0	False
```

**Comment by saeedmaleki** [comment]
I see, can you please share the entire NCCL log? Please enable `NCCL_DEBUG=INFO` and share the log. I think you have a different connection mechanism (non-P2P) which would be the difference.

**Comment by saeedmaleki** [comment]
One more thing to check: can you please try this branch? `fence_hack` I added a fence which could potentially be the necessary synchronization for non-P2P connections.

**Comment by liangyuRain** [comment]
Yes, I believe `fence_hack` has fixed the issue. Right now, it looks like the 8-channel schedules pass datacheck at all sizes. I will run more tests to verify.

**Comment by saeedmaleki** [comment]
OK phew. I think your connection requires proxy cpu threads and that's why a normal volatile load and store isn't enough and a fence is necessary. I need a bit more detail to know for sure.

**Comment by chughtapan** [comment]
Hi @saeedmaleki: by non P2P connections, do you mean that some two ranks i & j are communicating even without a direct network link between them? I don't believe that should be happening

Can you also explain a bit more about proxy cpu threads and when/how they are used? 

Thanks!

**Comment by saeedmaleki** [comment]
Hi Tapan,

There are three types of transports in MSCCL/NCCL: P2P (NVlink), shared memory (via host memory), and different NETs (IB, socket, etc.). You can find them [here](https://github.com/microsoft/msccl/tree/merging_with_nccl/src/transport). I believe your type of connection is a NET connection. With the exception of P2P transport (NVLink), every other connection needs a CPU proxy. You may follow the logic [here](https://github.com/microsoft/msccl/blob/217fcc8296682361d139c6712705acd4456c015d/src/init.cc#L839).

**Comment by chughtapan** [comment]
Thanks for the clarification! Yes, our connection is NET (IB)

**Comment by saeedmaleki** [comment]
@liangyuRain, if the fence has fixed your issues, I would make a PR and close this. Please let us know! Thanks again for these stress tests!

**Comment by liangyuRain** [comment]
@saeedmaleki, I have tested all the schedules I have, and the results look perfect. Thank you so much for the fixes!

**Comment by Eevan-zq** [comment]
@saeedmaleki Hello sir, I’m currently encountering the **'an illegal memory access was encountered'** issue as well, but I haven't found the **fence_hack** branch. Could you still help resolve this problem?

**Comment by liangyuRain** [comment]
> [@saeedmaleki](https://github.com/saeedmaleki) Hello sir, I’m currently encountering the **'an illegal memory access was encountered'** issue as well, but I haven't found the **fence_hack** branch. Could you still help resolve this problem?

As far as I have learned, [Azure/msccl](https://github.com/Azure/msccl) is the latest repo that has active maintenance.

**Comment by Eevan-zq** [comment]
Yes, I asked a question there, but no one replied.

**Comment by liangyuRain** [comment]
Then, based on my experience, it is highly likely some bug in the xml. Such as accessing offset outside of the declared i/o/s_chunks.

**Comment by Eevan-zq** [comment]
Okay, thank you. It seems that there will also be issues with the XML generated using msccl-tools.

**Comment by Eevan-zq** [comment]
By the way, I would also like to know how you tried the fence_hack branch, as I found that there is currently no such branch when I ran git branch -a. How did you execute it at that time?

**Comment by liangyuRain** [comment]
I suppose it existed back then. However, I can’t remember the specifics or find the branch because it was more than two years ago.

**Comment by Eevan-zq** [comment]
Yes, thank you for your reply.
